server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java

```diff
@@ -508,8 +532,97 @@ public ClusterMetadataManifest getLatestClusterMetadataManifest(String clusterNa
+    }
+
+    private Map<String, ClusterMetadataManifest> getLatestManifestForAllClusterUUIDs(String clusterName, Set<String> clusterUUIDs) {
+        Map<String, ClusterMetadataManifest> manifestsByClusterUUID = new HashMap<>();
+        for (String clusterUUID : clusterUUIDs) {
+            try {
+                ClusterMetadataManifest manifest = getLatestClusterMetadataManifest(clusterName, clusterUUID);
+                manifestsByClusterUUID.put(clusterUUID, manifest);
+            } catch (Exception e) {
+                throw new IllegalStateException(
+                    String.format(Locale.ROOT, "Exception in fetching manifest for clusterUUID: %s", clusterUUID)
+                );
+            }
+        }
+        return manifestsByClusterUUID;
+    }
+
+    /**
+     * This method creates a valid cluster UUID chain.
+     *
+     * @param manifestsByClusterUUID Map of latest ClusterMetadataManifest for every cluster UUID
+     * @return List of cluster UUIDs. The first element is the most recent cluster UUID in the chain
+     */
+    private List<String> createClusterChain(final Map<String, ClusterMetadataManifest> manifestsByClusterUUID) {
+        final Map<String, String> clusterUUIDGraph = manifestsByClusterUUID.values()
+            .stream()
+            .collect(Collectors.toMap(ClusterMetadataManifest::getClusterUUID, ClusterMetadataManifest::getPreviousClusterUUID));
+        final List<String> validClusterUUIDs = manifestsByClusterUUID.values()
+            .stream()
+            .filter(m -> !isInvalidClusterUUID(m) && !clusterUUIDGraph.containsValue(m.getClusterUUID()))
+            .map(ClusterMetadataManifest::getClusterUUID)
+            .collect(Collectors.toList());
+        if (validClusterUUIDs.isEmpty()) {
+            logger.info("There is no valid previous cluster UUID");
+            return Collections.emptyList();
+        }
+        if (validClusterUUIDs.size() > 1) {
+            throw new IllegalStateException(
+                String.format(
+                    Locale.ROOT,
+                    "The system has ended into multiple valid cluster states in the remote store. "
+                        + "Please check their latest manifest to decide which one you want to keep. Valid Cluster UUIDs: - %s",
+                    validClusterUUIDs
+                )
+            );
+        }
+        final List<String> validChain = new ArrayList<>();
+        String currentUUID = validClusterUUIDs.get(0);
+        while (!ClusterState.UNKNOWN_UUID.equals(currentUUID)) {
+            validChain.add(currentUUID);
+            // Getting the previous cluster UUID of a cluster UUID from the clusterUUID Graph
+            currentUUID = clusterUUIDGraph.get(currentUUID);
+        }
+        return validChain;
+    }
+
+    private boolean isInvalidClusterUUID(ClusterMetadataManifest manifest) {
+        return !manifest.isCommitted() && manifest.getIndices().isEmpty();
```

Comment:
why are we calling manifest with no indices as an invalid cluster ? would  the chain would break if we render it useless ?

Reply:
Manifest with no indices and NOT committed is invalid.
We publish state to remote before starting the publish phase. During this time another node would have published the state to remote causing a race condition and we may not be able to create a chain. So we reject the manifest which have no indices and are not committed.